docs/design/indexed-job.md

```diff
@@ -0,0 +1,891 @@
+run.  They will want to use existing images.  So, the image is not the place
+for the work list.
+
+A work list could be stored on networked storage.  However,
+unless the user knows the length of the list, the `.job.completions`
+cannot be set at the time of creating the job.  So, it would need
+to be copied somewhere else.  If it cannot be copied (e.g. because
+it changes often) then a different style of job, such as a work-queue Job
+should be used.
+
+A work list could be provided as immediate data in the Job Spec or in a
+volume such as `configData` referenced by the Job Spec.   We focus on these two
+options.
+
+### What Varies Between Pods of a Job
+
+Pods need to differ in some way to do something different. (They do not
+differ in the work-queue style of Job, but that style has ease-of-use issues).
+
+A general approach would be to allow pods to differ from each other in arbitrary ways.
+For example, the Job object could have a list of PodSpecs to run.
+However, this is so general that it provides little value.  It would:
+
+- make the Job Spec very verbose, especially for jobs with thousands of work items
+- Job becomes such a vague concept that it is hard to explain to users
+- in practice, we do not see cases where many pods which differ across many fields of their
+  specs, and need to run as a group, with no ordering constraints.
+- CLIs and UIs need to support more options for creating Job
+- it is useful for monitoring and accounting databases want to aggregate data for pods
+  with the same controller.  However, pods with very different Specs may not make sense
+  to aggregate.
+- profiling, debugging, accounting, auditing and monitoring tools cannot assume common
+  images/files, behaviors, provenance and so on between Pods of a Job.
+
+Also, variety has another cost.  Pods which differ in ways that affect scheduling
+(node constraints, resource requirements, labels) prevent the scheduler
+from treating them as fungible, which is an important optimization for the scheduler.
+
+Therefore, we will not allow Pods from the same Job to differ arbitrarily
+(anyway, users can use multiple Job objects for that case).  We will try to
+allow as little as possible to differ between pods of the same Job, while
+still allowing users to express common parallel patterns easily For users
+who need to run jobs which differ in other ways, they can create multiple
+Jobs, and manage them as a group using labels.
+
+
+From the above work lists, we see a need for Pods which differ in their command
+lines, and in their environment variables.  These work lists do not require the
+pods to differ in other ways.
+
+Experience in a [similar systems](http://static.googleusercontent.com/media/research.google.com/en//pubs/archive/43438.pdf) has shown this model to be applicable
+to a very broad range of problems, despite this restriction.
+
+Therefore we will allow pods in the same Job to differ **only** in the following aspects:
+
```

Comment:
Since indexed jobs are intended for static work distribution, and we have other controller for other use cases, I'm ok with the challenges completion args present to dynamic scalability.